### PR TITLE
[model] fix: align DeepSeek V4 runtime numerical semantics

### PR DESCRIPTION
**Pull request by Luosuu**
## Summary

Align the core DeepSeek-V4 runtime with official execution semantics before introducing quantization experiments.

- keep stateless prefill and training cache-free so the TileLang indexer remains reachable
- match official weighted RMSNorm multiplication order
- preserve checkpoint-stable RoPE dtypes in training and FP32 tables in inference
- normalize attention queries in the official BF16 order
- run top-k and hash router projections in FP32
- match the shared-expert FP32 clamped SwiGLU path

## Root cause

VeOmni created a DynamicCache even when use_cache was false, which forced the indexer onto its eager decode fallback. Several arithmetic-order differences in RMSNorm, routing, attention normalization, and shared experts also accumulated model-body divergence from official inference.

## Review scope

This is stack 1 of 3. It intentionally contains no FP8/FP4 inference kernels, QAT, or tracing scripts.

Tracking issue: #927
Original integration draft: #917

## Validation

- uv run pytest -q tests/models/test_deepseek_v4_fused_moe.py
- stateless-forward focused kernel test
- patchgen regeneration and dry-run
- git diff --check

## Stack navigation

Tracked by #927. Merge and review in this order:

1. #928 — runtime parity
2. #929 — FP8 activation QAT
3. #931 — parity diagnostics

#930 is closed as superseded because its inference-only FP4/GEMM paths are not used by training. The original #917 remains the integration and CI baseline; it is not the merge target for this stack.

### CODE REVIEW

File: veomni/models/transformers/deepseek_v4/deepseek_v4_gpu_patch_gen_config.py [hunk 1]
```diff
@@ -914,6 +968,67 @@ def _apply_gate(self, gate_up: torch.Tensor) -> torch.Tensor:
...
+@config.override_method(
+    "DeepseekV4TopKRouter.forward",
+    description="Match the official DeepSeek-V4 FP32 router projection",
+)
+def deepseek_v4_topk_router_forward_patched(
+    self,
+    hidden_states: torch.Tensor,
+) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    flat = hidden_states.reshape(-1, self.hidden_dim)
+    logits = F.linear(flat.float(), self.weight.float())
```
Comment (FoolPlayer) [inline veomni/models/transformers/deepseek_v4/deepseek_v4_gpu_patch_gen_config.py]: `F.linear` may be cast back to BF16 under an outer autocast context despite the explicit `.float()` calls. Please disable autocast around this projection; the same applies to `DeepseekV4HashRouter`, and an autocast-enabled regression test would help.

Comment (Luosuu) [reply veomni/models/transformers/deepseek_v4/deepseek_v4_gpu_patch_gen_config.py]: Fixed by disabling autocast around both TopK and HashRouter FP32 F.linear projections and adding an autocast-enabled CPU regression test covering both routers.

**Review by FoolPlayer — APPROVED** [review]
Reviewed the runtime parity changes, including stateless cache behavior, RMSNorm/RoPE dtype semantics, attention query normalization, FP32 router projections, shared-expert SwiGLU, patchgen consistency, and test coverage. No blocking issues found.
